File: lintcode/python/0071_binary_tree_zigzag_level_order_traversal.py
#!/usr/bin/python -t

# bfs, traversal way

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

from collections import deque

class Solution:
    """
    @param root: A Tree
    @return: A list of lists of integer include the zigzag level order traversal of its nodes' values.
    """
    def zigzagLevelOrder(self, root):
        # write your code here
        ret = []
        if root == None:
            return ret
            
        s = deque()
        s.append(root)
        cnt = 0
        
        while s:
            level = []
            q_len = len(s)
            
            for i in range(q_len):
                node = s.popleft()
                level.append(node.val)
                if node.left:
                    s.append(node.left)
                if node.right:
                    s.append(node.right)
                    
            if cnt %2 != 0:
                ret.append(level[::-1])
            else:
                ret.append(level)
                
            cnt += 1
                
        return ret
        


# dfs, level traversal way

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: A Tree
    @return: A list of lists of integer include the zigzag level order traversal of its nodes' values.
    """
    def zigzagLevelOrder(self, root):
        # write your code here
        self.ret = []
        self.preorder(root, 0, self.ret)
        
        return self.ret
        
    def preorder(self, node, level, ret):
        if node == None:
            return
        
        if len(ret) < level+1:
            ret.append([])
        if level %2 == 0:
            ret[level].append(node.val)
        else:
            ret[level].insert(0, node.val)
            
        self.preorder(node.left, level+1, ret)
        self.preorder(node.right, level+1, ret)
        
        return
    

File: lintcode/golang/0071_binary_tree_zigzag_level_order_traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

/**
 * @param root: A Tree
 * @return: A list of lists of integer include the zigzag level order traversal of its nodes' values.
 */
func zigzagLevelOrder (root *TreeNode) [][]int {
    // write your code here
    if root == nil {
        return [][]int{}
    }
    
    q := []*TreeNode{root}
    ret := [][]int{}
    needReverse := false
    
    for len(q) != 0 {
        level := []*TreeNode{}
        path := []int{}
        for _, node := range q {
            path = append(path, node.Val)
        
            if node.Left != nil {
                level = append(level, node.Left)
            }
            if node.Right != nil {
                level = append(level, node.Right)
            }
        }
        q = level
        if needReverse {
            reverse(path)
        }
        ret = append(ret, path)
        needReverse = !needReverse
    }
    
    return ret
}

func reverse (path []int) {
    l := 0
    r := len(path)-1
    
    for l < r {
        tmp := path[l]
        path[l] = path[r]
        path[r] = tmp
        l++
        r--
    }
    
    return
}
